File: core/src/main/java/com/avast/continuity/javaapi/Continuity.java
package com.avast.continuity.javaapi;

import com.avast.continuity.Continuity$;
import com.avast.continuity.IdentityThreadNamer$;
import com.avast.continuity.ThreadNamer;
import scala.Option;
import scala.runtime.AbstractFunction0;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.Callable;
import java.util.concurrent.Executor;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.ScheduledExecutorService;

/* KEEP THIS IMPLEMENTATION IN SYNC WITH THE SCALA API VERSION */

/**
 * Provides methods to work with the Continuity context and factory methods to wrap thread pools.
 */
public final class Continuity {

    private static final Continuity$ CONTINUITY = Continuity$.MODULE$;

    /**
     * Marks that the current thread was already processed.
     */
    private static final String MARKER = "__MARKER__";

    private Continuity() {
    }

    /**
     * <p>Puts the given values into the context, names a thread and runs the given block of code.
     * It correctly cleans up everything after the block finishes.</p>
     * <p>This method is to be used at the leaves of Continuity context usage meaning that you have to fill in the context
     * somewhere and this method should be used for that. From there the context is propagated automatically.</p>
     */
    public static <T> T withContext(Map<String, String> ctxValues, Callable<T> block) throws RuntimeException {
        return withContext(ctxValues, IdentityThreadNamer$.MODULE$, block);
    }

    /**
     * <p>Puts the given values into the context, names a thread and runs the given block of code.
     * It correctly cleans up everything after the block finishes.</p>
     * <p>This method is to be used at the leaves of Continuity context usage meaning that you have to fill in the context
     * somewhere and this method should be used for that. From there the context is propagated automatically.</p>
     */
    public static <T> T withContext(Map<String, String> ctxValues, ThreadNamer threadNamer, Callable<T> block) throws RuntimeException {
        if (getFromContext(MARKER).isPresent()) {
            try {
                return block.call();
            } catch (Exception ex) {
                throw new RuntimeException(ex);
            }
        } else {
            try {
                putToContext(MARKER, "");
                ctxValues.forEach(Continuity::putToContext);
                return threadNamer.nameThread(new AbstractFunction0<T>() {
                    @Override
                    public T apply() {
                        try {
                            return block.call();
                        } catch (Exception ex) {
                            throw new RuntimeException(ex);
                        }
                    }
                });
            } finally {
                ctxValues.forEach((k, v) -> removeFromContext(k));
                removeFromContext(MARKER);
            }
        }
    }

    public static Optional<String> getFromContext(String key) {
        return optionToOptional(CONTINUITY.getFromContext(key));
    }

    public static void putToContext(String key, String value) {
        CONTINUITY.putToContext(key, value);
    }

    public static void removeFromContext(String key) {
        CONTINUITY.removeFromContext(key);
    }

    public static Executor wrapExecutor(Executor executor) {
        return CONTINUITY.wrapExecutor(executor, IdentityThreadNamer$.MODULE$);
    }

    public static Executor wrapExecutor(Executor executor, ThreadNamer threadNamer) {
        return CONTINUITY.wrapExecutor(executor, threadNamer);
    }

    public static ExecutorService wrapExecutorService(ExecutorService executor) {
        return CONTINUITY.wrapExecutorService(executor, IdentityThreadNamer$.MODULE$);
    }

    public static ExecutorService wrapExecutorService(ExecutorService executor, ThreadNamer threadNamer) {
        return CONTINUITY.wrapExecutorService(executor, threadNamer);
    }

    public static ScheduledExecutorService wrapScheduledExecutorService(ScheduledExecutorService executor) {
        return CONTINUITY.wrapScheduledExecutorService(executor, IdentityThreadNamer$.MODULE$);
    }

    public static ScheduledExecutorService wrapScheduledExecutorService(ScheduledExecutorService executor, ThreadNamer threadNamer) {
        return CONTINUITY.wrapScheduledExecutorService(executor, threadNamer);
    }

    private static <T> Optional<T> optionToOptional(Option<T> option) {
        if (option.isDefined()) {
            return Optional.of(option.get());
        } else {
            return Optional.empty();
        }
    }

}

File: core-scala/src/main/scala/com/avast/continuity/Continuity.scala
package com.avast.continuity

import java.util.concurrent.{Executor, ExecutorService, ScheduledExecutorService}

import org.slf4j.MDC

import scala.concurrent.{ExecutionContext, ExecutionContextExecutor, ExecutionContextExecutorService}

/* KEEP THIS IMPLEMENTATION IN SYNC WITH THE JAVA API VERSION */

trait Continuity {

  protected[this] def withContext[A](ctxValues: (String, String)*)(block: => A)(
      implicit threadNamer: ThreadNamer = IdentityThreadNamer): A = {
    Continuity.withContext(ctxValues: _*)(block)
  }

  protected[this] def getFromContext(key: String): Option[String] = Continuity.getFromContext(key)

  protected[this] def putToContext(key: String, value: String): Unit = Continuity.putToContext(key, value)

  protected[this] def removeFromContext(key: String): Unit = Continuity.removeFromContext(key)

}

/** Provides methods to work with the Continuity context and factory methods to wrap thread pools. */
object Continuity {

  /** Marks that the current thread was already processed. */
  private final val Marker = "__MARKER__"

  /** Creates an instance of [[com.avast.continuity.Continuity]] in case you don't want to use the static methods. */
  def apply(implicit threadNamer: ThreadNamer = IdentityThreadNamer): Continuity = new Continuity {}

  /** Puts the given values into the context, names a thread and runs the given block of code.
    * It correctly cleans up everything after the block finishes.
    *
    * This method is to be used at the leaves of Continuity context usage meaning that you have to fill in the context
    * somewhere and this method should be used for that. From there the context is propagated automatically.
    */
  def withContext[A](ctxValues: (String, String)*)(block: => A)(implicit threadNamer: ThreadNamer = IdentityThreadNamer): A = {
    if (getFromContext(Marker).isEmpty) {
      try {
        putToContext(Marker, "")
        ctxValues.foreach {
          case (key, value) =>
            putToContext(key, value)
        }

        threadNamer.nameThread {
          block
        }
      } finally {
        ctxValues.foreach {
          case (key, _) =>
            removeFromContext(key)
        }
        removeFromContext(Marker)
      }
    } else {
      block
    }
  }

  def getFromContext(key: String): Option[String] = context.get.get(key)

  def putToContext(key: String, value: String): Unit = {
    val ctx = context.get
    context.set(ctx + (key -> value))
    MDC.put(key, value)
  }

  def removeFromContext(key: String): Unit = {
    val ctx = context.get
    context.set(ctx - key)
    MDC.remove(key)
  }

  def wrapExecutionContext(executor: ExecutionContext)(implicit threadNamer: ThreadNamer = IdentityThreadNamer): ExecutionContext =
    preventDoubleWrap(executor) {
      new ContinuityExecutionContext(executor)(threadNamer)
    }

  def wrapExecutionContextExecutor(executor: ExecutionContextExecutor)(
      implicit threadNamer: ThreadNamer = IdentityThreadNamer): ExecutionContextExecutor = preventDoubleWrap(executor) {
    new ContinuityExecutionContextExecutor(executor)(threadNamer)
  }

  def wrapExecutionContextExecutorService(executor: ExecutionContextExecutorService)(
      implicit threadNamer: ThreadNamer = IdentityThreadNamer): ExecutionContextExecutorService = {
    preventDoubleWrap(executor) {
      new ContinuityExecutionContextExecutorService(executor)(threadNamer)
    }
  }

  def wrapExecutor(executor: Executor)(implicit threadNamer: ThreadNamer = IdentityThreadNamer): Executor = preventDoubleWrap(executor) {
    new ContinuityExecutor(executor)(threadNamer)
  }

  def wrapExecutorService(executor: ExecutorService)(implicit threadNamer: ThreadNamer = IdentityThreadNamer): ExecutorService =
    preventDoubleWrap(executor) {
      new ContinuityExecutorService(executor)(threadNamer)
    }

  def wrapScheduledExecutorService(executor: ScheduledExecutorService)(
      implicit threadNamer: ThreadNamer = IdentityThreadNamer): ScheduledExecutorService =
    preventDoubleWrap(executor) {
      new ContinuityScheduledExecutorService(executor)(threadNamer)
    }

  private def preventDoubleWrap[A](executor: A)(wrap: => A): A = executor match {
    case _: ContinuityExecutorMarker => executor
    case _ => wrap
  }

}
